File: БР1.2/Шестак_Богдан/labs/lab3/application-service/app/publisher.py
import aio_pika
import json
import os
import logging

RABBITMQ_URL = os.getenv("RABBITMQ_URL", "amqp://guest:guest@rabbitmq:5672/")

logger = logging.getLogger(__name__)


async def get_connection() -> aio_pika.abc.AbstractRobustConnection:
    return await aio_pika.connect_robust(RABBITMQ_URL)


async def publish_event(exchange_name: str, routing_key: str, payload: dict):
    """
    Publishes a JSON event to the given exchange with the given routing_key.
    Uses topic exchange so consumers can bind with patterns.
    """
    try:
        connection = await get_connection()
        async with connection:
            channel = await connection.channel()
            exchange = await channel.declare_exchange(
                exchange_name,
                aio_pika.ExchangeType.TOPIC,
                durable=True,
            )
            message = aio_pika.Message(
                body=json.dumps(payload).encode(),
                content_type="application/json",
                delivery_mode=aio_pika.DeliveryMode.PERSISTENT,
            )
            await exchange.publish(message, routing_key=routing_key)
            logger.info(f"Published event [{routing_key}] to [{exchange_name}]: {payload}")
    except Exception as e:
        # Non-critical: log and continue — don't fail the main request
        logger.error(f"Failed to publish event [{routing_key}]: {e}")

File: БР1.2/Москалец Данила/labs/lab2/sangel-microservices/order-service/event-bus/src/publisher.ts
import { getChannel } from './connection';
import { Exchanges } from './events';

// Объявляем exchanges при старте
export async function setupExchanges(): Promise<void> {
  const channel = await getChannel();
  
  for (const exchange of Object.values(Exchanges)) {
    await channel.assertExchange(exchange, 'topic', { durable: true });
    console.log(`Exchange declared: ${exchange}`);
  }
}

export async function publishEvent(
  exchange: string,
  routingKey: string,
  data: any
): Promise<void> {
  try {
    const channel = await getChannel();
    const message = Buffer.from(JSON.stringify({
      ...data,
      timestamp: data.timestamp || new Date().toISOString(),
    }));
    
    channel.publish(exchange, routingKey, message, { persistent: true });
    console.log(`📤 Event published: ${routingKey}`, data);
  } catch (error) {
    console.error(`Failed to publish event ${routingKey}:`, error);
  }
}